aws/resource_aws_waf_regex_pattern_set.go

```diff
@@ -0,0 +1,176 @@
+		Create: resourceAwsWafRegexPatternSetCreate,
+		Read:   resourceAwsWafRegexPatternSetRead,
+		Update: resourceAwsWafRegexPatternSetUpdate,
+		Delete: resourceAwsWafRegexPatternSetDelete,
+
+		Schema: map[string]*schema.Schema{
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"regex_pattern_strings": &schema.Schema{
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+		},
+	}
+}
+
+func resourceAwsWafRegexPatternSetCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).wafconn
+
+	log.Printf("[INFO] Creating WAF Regex Pattern Set: %s", d.Get("name").(string))
+
+	wr := newWafRetryer(conn, "global")
+	out, err := wr.RetryWithToken(func(token *string) (interface{}, error) {
+		params := &waf.CreateRegexPatternSetInput{
+			ChangeToken: token,
+			Name:        aws.String(d.Get("name").(string)),
+		}
+		return conn.CreateRegexPatternSet(params)
+	})
+	if err != nil {
+		return fmt.Errorf("Failed creating WAF Regex Pattern Set: %s", err)
+	}
+	resp := out.(*waf.CreateRegexPatternSetOutput)
+
+	d.SetId(*resp.RegexPatternSet.RegexPatternSetId)
+
+	return resourceAwsWafRegexPatternSetUpdate(d, meta)
+}
+
+func resourceAwsWafRegexPatternSetRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).wafconn
+	log.Printf("[INFO] Reading WAF Regex Pattern Set: %s", d.Get("name").(string))
+	params := &waf.GetRegexPatternSetInput{
+		RegexPatternSetId: aws.String(d.Id()),
+	}
+
+	resp, err := conn.GetRegexPatternSet(params)
+	if err != nil {
+		if isAWSErr(err, "WAFNonexistentItemException", "") {
+			log.Printf("[WARN] WAF Regex Pattern Set (%s) not found, removing from state", d.Id())
+			d.SetId("")
+			return nil
+		}
+
+		return err
+	}
+
+	d.Set("name", resp.RegexPatternSet.Name)
+	d.Set("regex_pattern_strings", aws.StringValueSlice(resp.RegexPatternSet.RegexPatternStrings))
+
+	return nil
+}
+
+func resourceAwsWafRegexPatternSetUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).wafconn
+
+	log.Printf("[INFO] Updating WAF Regex Pattern Set: %s", d.Get("name").(string))
+
+	if d.HasChange("regex_pattern_strings") {
+		o, n := d.GetChange("regex_pattern_strings")
+		oldPatterns, newPatterns := o.(*schema.Set).List(), n.(*schema.Set).List()
+		err := updateWafRegexPatternSetPatternStrings(d.Id(), oldPatterns, newPatterns, conn)
+		if err != nil {
+			return fmt.Errorf("Failed updating WAF Regex Pattern Set: %s", err)
+		}
+	}
+
+	return resourceAwsWafRegexPatternSetRead(d, meta)
+}
+
+func resourceAwsWafRegexPatternSetDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).wafconn
+
+	oldPatterns := d.Get("regex_pattern_strings").(*schema.Set).List()
+	if len(oldPatterns) > 0 {
+		noPatterns := []interface{}{}
+		err := updateWafRegexPatternSetPatternStrings(d.Id(), oldPatterns, noPatterns, conn)
```

Comment:
Out of my own curiosity, what error does it throw if you do not empty the patterns before attempting to delete? Maybe we should include it as a comment?